README
Orthanc - A Lightweight, RESTful DICOM Server
=============================================


General Information
-------------------

This repository contains an ImageJ plugin that allows to easily access
DICOM images that are stored in some Orthanc server. This plugin might
be useful to medical physicists for QA tasks, as well as for
pedagogical purpose.

Demonstration: http://youtu.be/wsslVhE-R1Q

The instructions for building/installing the plugin can be found in
the "INSTALL" file.


Licensing
---------

Orthanc and this plugin are licensed under the GPLv3 license.

We also kindly ask scientific works and clinical studies that make use
of this plugin or of Orthanc, to cite Orthanc in their associated
publications. Similarly, we ask open-source and closed-source products
that make use of Orthanc to warn us about this use. You can cite our
work using the following BibTeX entry:

@inproceedings{Jodogne:ISBI2013,
  author = {Jodogne, S. and Bernard, C. and Devillers, M. and Lenaerts, E. and Coucke, P.},
  title = {Orthanc -- {A} Lightweight, {REST}ful {DICOM} Server for Healthcare and Medical Research},
  booktitle={Biomedical Imaging ({ISBI}), {IEEE} 10th International Symposium on}, 
  year={2013}, 
  pages={190-193}, 
  ISSN={1945-7928},
  month=apr,
  url={http://ieeexplore.ieee.org/xpl/articleDetails.jsp?tp=&arnumber=6556444},
  address={San Francisco, {CA}, {USA}}
}


Disclaimer
----------

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


Dependencies
------------

This plugin depends upon the "json-simple" toolkit
(http://code.google.com/p/json-simple/), that is released under the
Apache License 2.0. The source of this dependency is located in the
"org/" folder.


NEWS
Pending changes in the mainline
===============================



Version 1.0.0 (2014/12/08)
==========================

* Initial release


com/orthancserver/Orthanc.java
/**
 * Orthanc - A Lightweight, RESTful DICOM Store
 * Copyright (C) 2012-2014 Medical Physics Department, CHU of Liege,
 * Belgium
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 **/


package com.orthancserver;

import ij.plugin.PlugIn;
import ij.Prefs;
import javax.swing.JOptionPane;

public class Orthanc implements PlugIn 
{
  public void run(String arg) 
  {
    SelectImageDialog d = new SelectImageDialog();
    d.Unserialize(Prefs.get("orthanc.servers", ""));

    boolean success = d.ShowModal();
    Prefs.set("orthanc.servers", d.Serialize());  

    if (success)
    {
      try 
      {
        DicomDecoder decoder = new DicomDecoder(d.GetSelectedConnection(), d.IsInstanceSelected(), d.GetSelectedUuid());
        decoder.GetImage().show();
      }
      catch (Exception e) 
      {
        JOptionPane.showMessageDialog(null, "Error while importing this image: " + e.getMessage());
        //e.printStackTrace(System.out);
      }
    }
  }
}


com/orthancserver/GridLayout2.java
/**
 * Orthanc - A Lightweight, RESTful DICOM Store
 * Copyright (C) 2012-2014 Medical Physics Department, CHU of Liege,
 * Belgium
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 **/


package com.orthancserver;

import java.awt.*;

public class GridLayout2 extends GridLayout 
{
  public GridLayout2() {
    this(1, 0, 0, 0);
  }

  public GridLayout2(int rows, int cols) {
    this(rows, cols, 0, 0);
  }

  public GridLayout2(int rows, int cols, int hgap, int vgap) {
    super(rows, cols, hgap, vgap);
  }

  public Dimension preferredLayoutSize(Container parent) {
    synchronized (parent.getTreeLock()) {
      Insets insets = parent.getInsets();
      int ncomponents = parent.getComponentCount();
      int nrows = getRows();
      int ncols = getColumns();
      if (nrows > 0) {
        ncols = (ncomponents + nrows - 1) / nrows;
      } 
      else {
        nrows = (ncomponents + ncols - 1) / ncols;
      }
      int[] w = new int[ncols];
      int[] h = new int[nrows];
      for (int i = 0; i < ncomponents; i ++) {
        int r = i / ncols;
        int c = i % ncols;
        Component comp = parent.getComponent(i);
        Dimension d = comp.getPreferredSize();
        if (w[c] < d.width) {
          w[c] = d.width;
        }
        if (h[r] < d.height) {
          h[r] = d.height;
        }
      }
      int nw = 0;
      for (int j = 0; j < ncols; j ++) {
        nw += w[j];
      }
      int nh = 0;
      for (int i = 0; i < nrows; i ++) {
        nh += h[i];
      }
      return new Dimension(insets.left + insets.right + nw + (ncols-1)*getHgap(), 
                           insets.top + insets.bottom + nh + (nrows-1)*getVgap());
    }
  }

  public Dimension minimumLayoutSize(Container parent) {
    synchronized (parent.getTreeLock()) {
      Insets insets = parent.getInsets();
      int ncomponents = parent.getComponentCount();
      int nrows = getRows();
      int ncols = getColumns();
      if (nrows > 0) {
        ncols = (ncomponents + nrows - 1) / nrows;
      } 
      else {
        nrows = (ncomponents + ncols - 1) / ncols;
      }
      int[] w = new int[ncols];
      int[] h = new int[nrows];
      for (int i = 0; i < ncomponents; i ++) {
        int r = i / ncols;
        int c = i % ncols;
        Component comp = parent.getComponent(i);
        Dimension d = comp.getMinimumSize();
        if (w[c] < d.width) {
          w[c] = d.width;
        }
        if (h[r] < d.height) {
          h[r] = d.height;
        }
      }
      int nw = 0;
      for (int j = 0; j < ncols; j ++) {
        nw += w[j];
      }
      int nh = 0;
      for (int i = 0; i < nrows; i ++) {
        nh += h[i];
      }
      return new Dimension(insets.left + insets.right + nw + (ncols-1)*getHgap(), 
                           insets.top + insets.bottom + nh + (nrows-1)*getVgap());
    }
  }

  public void layoutContainer(Container parent) {
    synchronized (parent.getTreeLock()) {
      Insets insets = parent.getInsets();
      int ncomponents = parent.getComponentCount();
      int nrows = getRows();
      int ncols = getColumns();
      if (ncomponents == 0) {
        return;
      }
      if (nrows > 0) {
        ncols = (ncomponents + nrows - 1) / nrows;
      } 
      else {
        nrows = (ncomponents + ncols - 1) / ncols;
      }
      int hgap = getHgap();
      int vgap = getVgap();
      // scaling factors      
      Dimension pd = preferredLayoutSize(parent);
      double sw = (1.0 * parent.getWidth()) / pd.width;
      double sh = (1.0 * parent.getHeight()) / pd.height;
      // scale
      int[] w = new int[ncols];
      int[] h = new int[nrows];
      for (int i = 0; i < ncomponents; i ++) {
        int r = i / ncols;
        int c = i % ncols;
        Component comp = parent.getComponent(i);
        Dimension d = comp.getPreferredSize();
        d.width = (int) (sw * d.width);
        d.height = (int) (sh * d.height);
        if (w[c] < d.width) {
          w[c] = d.width;
        }
        if (h[r] < d.height) {
          h[r] = d.height;
        }
      }
      for (int c = 0, x = insets.left; c < ncols; c ++) {
        for (int r = 0, y = insets.top; r < nrows; r ++) {
          int i = r * ncols + c;
          if (i < ncomponents) {
            parent.getComponent(i).setBounds(x, y, w[c], h[r]);
          }
          y += h[r] + vgap;
        }
        x += w[c] + hgap;
      }
    }
  }  
}

com/orthancserver/DicomDecoder.java
/**
 * Orthanc - A Lightweight, RESTful DICOM Store
 * Copyright (C) 2012-2014 Medical Physics Department, CHU of Liege,
 * Belgium
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 **/


package com.orthancserver;

import ij.ImagePlus;
import ij.ImageStack;
import ij.process.ImageProcessor;
import ij.process.ShortProcessor;
import ij.process.ColorProcessor;
import ij.io.FileInfo;
import ij.measure.Calibration;
import org.json.simple.*;
import java.util.List;
import java.util.ArrayList;
import java.util.Collections;
import java.io.IOException;
import javax.swing.SwingWorker;
import java.util.concurrent.ExecutionException;

import javax.swing.JFrame;
import javax.swing.JProgressBar;
import java.awt.BorderLayout;
import javax.swing.JPanel;
import javax.swing.JButton;
import java.awt.FlowLayout;
import javax.swing.border.EmptyBorder;
import java.awt.event.ActionListener;
import java.awt.event.ActionEvent;


public class DicomDecoder
{
  private static class ProgressDialog extends JFrame
  {
    private JProgressBar bar_ = new JProgressBar();
    private boolean canceled_ = false;
    
    public ProgressDialog(int count)
    {
      getContentPane().setLayout(new BorderLayout());
      bar_.setBorder(new EmptyBorder(20, 20, 20, 20));
      getContentPane().add(bar_, BorderLayout.NORTH);

      JPanel buttonPane = new JPanel();
			buttonPane.setLayout(new FlowLayout(FlowLayout.RIGHT));
			getContentPane().add(buttonPane, BorderLayout.SOUTH);

      JButton cancelButton = new JButton("Cancel");
      cancelButton.addActionListener(new ActionListener() {
        public void actionPerformed(ActionEvent arg) {
          canceled_ = true;
        }
      });
      buttonPane.add(cancelButton);

      bar_.setMaximum(0);
      bar_.setMaximum(count);

      setSize(500, 150);
      setTitle("Importing series from Orthanc");
      setLocationRelativeTo(null);  // Center dialog on screen
    }

    public void SetProgress(int value)
    {
      bar_.setValue(value);
    }

    public boolean IsCanceled()
    {
      return canceled_;
    }
  };

  private static void ExtractCalibration(ImagePlus image,
                                         JSONObject tags)
  {
    JSONObject rescaleIntercept = (JSONObject) tags.get("0028,1052");
    JSONObject rescaleSlope = (JSONObject) tags.get("0028,1053");
    if (rescaleIntercept != null &&
        rescaleSlope != null)
    {
      double[] coeff = {
        Float.valueOf((String) rescaleIntercept.get("Value")),
        Float.valueOf((String) rescaleSlope.get("Value"))
      };
      image.getCalibration().setFunction(Calibration.STRAIGHT_LINE, coeff, "Gray Value");
    }
  }

  private static void ExtractPixelSpacing(ImagePlus image,
                                          JSONObject tags)
  {
    JSONObject pixelSpacing = (JSONObject) tags.get("0028,0030");
    if (pixelSpacing != null)
    {
      String[] tokens = ((String) pixelSpacing.get("Value")).split("\\\\");
      if (tokens.length == 2)
      {
        FileInfo fi = image.getFileInfo();
        fi.pixelWidth = Float.valueOf(tokens[0]);
        fi.pixelHeight = Float.valueOf(tokens[1]);
        fi.unit = "mm";

        image.setFileInfo(fi);
        image.getCalibration().pixelWidth = fi.pixelWidth;
        image.getCalibration().pixelHeight = fi.pixelHeight;
        image.getCalibration().setUnit(fi.unit);
      }
    }
  }

  private static void ExtractDicomInfo(ImagePlus image,
                                       JSONObject tags)
  {
    String info = new String();

    ArrayList<String> tagsIndex = new ArrayList<String>();
    for (Object tag : tags.keySet())
    {
      tagsIndex.add((String) tag);
    }

    Collections.sort(tagsIndex);
    for (String tag : tagsIndex) 
    {
      JSONObject value = (JSONObject) tags.get(tag);

      if (((String) value.get("Type")).equals("String"))
      {
        info += (tag + " " + (String) value.get("Name") +
                 ": " + (String) value.get("Value") + "\n");
      }
    }

    image.setProperty("Info", info);
  }

  private static ImageProcessor DecodeInstance(OrthancConnection c,
                                               String uuid) throws IOException
  {
    try
    {
      String uri = "/instances/" + uuid + "/image-uint16";
      ShortProcessor slice = new ShortProcessor(c.ReadImage(uri));
      return slice;
    }
    catch (IllegalArgumentException e)
    {
      // Color image
      String uri = "/instances/" + uuid + "/preview";
      ColorProcessor slice = new ColorProcessor(c.ReadImage(uri));
      return slice;
    }
  }

  private static ImageStack AddSlice(ImageStack stack,
                                     OrthancConnection c,
                                     String uuid) throws IOException
  {
    ImageProcessor slice = DecodeInstance(c, uuid);

    if (stack == null)
    {
      stack = new ImageStack(slice.getWidth(), slice.getHeight());
    }

    stack.addSlice(slice);
    return stack;
  }


  static private class Slice implements Comparable
  {
    private Float index_;
    private String uuid_;
    
    Slice(float index,
          String uuid)
    {
      index_ = index;
      uuid_ = uuid;
    }

		@Override
    public int compareTo(Object other) 
		{
			return index_.compareTo(((Slice) other).index_);
    }

    public String GetUuid()
    {
      return uuid_;
    }
  }



  private String[] SortSlices(List<Slice> slices)
  {
    Collections.sort(slices);

    String[] result = new String[slices.size()];

    for (int i = 0; i < slices.size(); i++)
    {
      result[i] = slices.get(i).GetUuid();
    }

    return result;
  }



  private String[]  SortSlicesBy3D(OrthancConnection c, 
                                   JSONArray instances) throws IOException
  {
    ArrayList<Slice> slices = new ArrayList<Slice>();
    float normal[] = null;

    float minDistance = Float.POSITIVE_INFINITY;
    float maxDistance = Float.NEGATIVE_INFINITY;

    for (int i = 0; i < instances.size(); i++)
    {
      String uuid = (String) instances.get(i);
      JSONObject instance = (JSONObject) c.ReadJson("/instances/" + uuid + "/tags?simplify");
      if (!instance.containsKey("ImageOrientationPatient") ||
          !instance.containsKey("ImagePositionPatient"))
      {
        return null;
      }

      if (i == 0)
      {
        String[] tokens = ((String) instance.get("ImageOrientationPatient")).split("\\\\");
        if (tokens.length != 6)
        {
          return null;
        }

        float cosines[] = new float[6];
        for (int j = 0; j < 6; j++)
        {
          cosines[j] = Float.parseFloat(tokens[j]);
        }

        normal = new float[] {
          cosines[1] * cosines[5] - cosines[2] * cosines[4],
          cosines[2] * cosines[3] - cosines[0] * cosines[5],
          cosines[0] * cosines[4] - cosines[1] * cosines[3]
        };
      }

      String[] tokens = ((String) instance.get("ImagePositionPatient")).split("\\\\");
      if (tokens.length != 3)
      {
        return null;
      }

      float distance = 0;
      for (int j = 0; j < 3; j++)
      {
        distance += normal[j] * Float.parseFloat(tokens[j]);
      }

      minDistance = Math.min(minDistance, distance);
      maxDistance = Math.max(minDistance, distance);
      slices.add(new Slice(distance, uuid));
    }

    if (maxDistance - minDistance < 0.001)
    {
      return null;
    }

    return SortSlices(slices);
  }


  private String[]  SortSlicesByNumber(OrthancConnection c, 
                                       JSONArray instances) throws IOException
  {
    ArrayList<Slice> slices = new ArrayList<Slice>();

    for (int i = 0; i < instances.size(); i++)
    {
      String uuid = (String) instances.get(i);
      JSONObject instance = (JSONObject) c.ReadJson("/instances/" + uuid);
      Long index = (Long) instance.get("IndexInSeries");
      slices.add(new Slice((float) index, uuid));
    }

    return SortSlices(slices);
  }



  private String[] GetSlices(OrthancConnection c, 
                             JSONArray instances) throws IOException
  {
    String[] result;

    result = SortSlicesBy3D(c, instances);
    if (result != null && result.length == instances.size())
    {
      return result;
    }

    result = SortSlicesByNumber(c, instances);
    if (result != null && result.length == instances.size())
    {
      return result;
    }

    throw new IOException("Not a 3D image");
  }




  private ImagePlus image_;

  public DicomDecoder(final OrthancConnection c,
                      boolean isInstance,
                      String uuid) throws IOException, InterruptedException, ExecutionException
  {
    ImageStack stack = null;
    JSONObject tags = null;
    String tagsUri, name;

    if (isInstance)
    {
      name = "Instance " + uuid;
      tags = (JSONObject) c.ReadJson("/instances/" + uuid + "/tags");
      stack = AddSlice(stack, c, uuid);
    }
    else
    {
      name = "Series " + uuid;

      JSONObject series = (JSONObject) c.ReadJson("/series/" + uuid);
      JSONArray instances = (JSONArray) series.get("Instances");

      try
      {
        tags = (JSONObject) c.ReadJson("/series/" + uuid + "/shared-tags");
      }
      catch (Exception e)
      {
        // Fallback for old versions of Orthanc, without "shared-tags"
        tags = (JSONObject) c.ReadJson("/instances/" + (String) instances.get(0) + "/tags");
      }

      final String[] slices = GetSlices(c, instances);
      final ProgressDialog progress = new ProgressDialog(slices.length);

      try
      {
        progress.setVisible(true);
        SwingWorker<ImageStack, Float> worker = new SwingWorker<ImageStack, Float>() {
          @Override
          protected ImageStack doInBackground()
          {
            try
            {
              ImageStack stack = null;

              for (int i = 0; i < slices.length; i++)
              {
                if (progress.IsCanceled())
                {
                  return null;
                }

                progress.SetProgress(i);
                stack = AddSlice(stack, c, slices[i]);
              }

              return stack;
            }
            catch (IOException e)
            {
              return null;
            }
          }
        };

        worker.execute();
        stack = worker.get();
      }
      finally
      {
        progress.setVisible(false);
      }
    }

    image_ = new ImagePlus(name, stack);
    
    ExtractCalibration(image_, tags);
    ExtractPixelSpacing(image_, tags);
    ExtractDicomInfo(image_, tags);
  }

  public ImagePlus GetImage()
  {
    return image_;
  }
}


com/orthancserver/PreviewPanel.java
/**
 * Orthanc - A Lightweight, RESTful DICOM Store
 * Copyright (C) 2012-2014 Medical Physics Department, CHU of Liege,
 * Belgium
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 **/


package com.orthancserver;

import java.awt.Graphics;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.ByteArrayInputStream;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.imageio.ImageIO;
import javax.swing.JPanel;

public class PreviewPanel extends JPanel
{
  private BufferedImage image_ = null;

  public void Reset()
  {
    image_ = null;
    repaint();
  }

  public void Load(OrthancConnection orthanc,
                   String uri)
  {
    try 
    {         
      image_ = orthanc.ReadImage(uri);
      repaint();
    }
    catch (IOException e) 
    {
      Reset();
    }
  }

  @Override
  protected void paintComponent(Graphics g) 
  {
    super.paintComponent(g);

    if (image_ != null)
    {
      float scaleX = (float) getWidth() / (float) image_.getWidth();
      float scaleY = (float) getHeight() / (float) image_.getHeight();
      float scale = Math.min(scaleX, scaleY);
      if (scale > 1)
      {
        // Do not upscale the image
        scale = 1;
      }

      int width = Math.round((float) image_.getWidth() * scale);
      int height = Math.round((float) image_.getHeight() * scale);

      int dx = (getWidth() - width) / 2;
      int dy = (getHeight() - height) / 2;
      g.drawImage(image_, dx, dy, dx + width, dy + height, 
                  0, 0, image_.getWidth(), image_.getHeight(), null);
    }
  }
}


com/orthancserver/OrthancConnection.java
/**
 * Orthanc - A Lightweight, RESTful DICOM Store
 * Copyright (C) 2012-2014 Medical Physics Department, CHU of Liege,
 * Belgium
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 **/


package com.orthancserver;

import java.awt.image.BufferedImage;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.ConnectException;
import java.net.URL;
import java.net.URLConnection;
import javax.imageio.ImageIO;
import javax.xml.bind.DatatypeConverter;
import org.json.simple.JSONObject;
import org.json.simple.JSONValue;

public class OrthancConnection
{
	private String baseUrl_;
	private String authentication_;
  private String name_;
	
	OrthancConnection()
	{
		baseUrl_ = "http://localhost:8042/";
		authentication_ = null;
    name_ = "Default Orthanc server";
	}


  private static InputStream OpenUrl(String urlString) throws IOException
	{
		URL url = new URL(urlString);

    try
    {
      return url.openStream();
    }
    catch (ConnectException e)
    {
      throw new IOException();
    }
	}

	
	private static InputStream OpenUrl(String urlString,
																		 String authentication) throws IOException
	{
    try
    {
      URL url = new URL(urlString);
      // http://blogs.deepal.org/2008/01/sending-basic-authentication-using-url.html
      URLConnection uc = url.openConnection();
      uc.setRequestProperty("Authorization", "Basic " + authentication);
      return uc.getInputStream();
    }
    catch (ConnectException e)
    {
      throw new IOException();
    }
	}


  private static String ComputeAuthentication(String username,
                                              String password)
  {
		String auth = (new StringBuffer(username).append(":").append(password)).toString();
		// http://stackoverflow.com/a/14413290/881731
		return DatatypeConverter.printBase64Binary(auth.getBytes());
  }

  
  public InputStream OpenStream(String uri) throws IOException
  {
    String url = baseUrl_ + uri;
    return authentication_ == null ? OpenUrl(url) : OpenUrl(url, authentication_);
  }


	public void SetCredentials(String username,
                             String password)
	{
		authentication_ = ComputeAuthentication(username, password);
	}

  public void SetBaseUrl(String url)
  {
    if (url.endsWith("/"))
    {
      baseUrl_ = url;
    }
    else
    {
      baseUrl_ = url + "/";
    }
  }

  public String GetBaseUrl()
  {
    return baseUrl_;
  }

  public String ReadString(String uri) throws IOException
  {
    InputStream stream = OpenStream(uri);

    BufferedReader reader = null;
    try 
		{
			reader = new BufferedReader(new InputStreamReader(stream));

      StringBuffer buffer = new StringBuffer();
      int read;
      char[] chars = new char[1024 * 16];
      while ((read = reader.read(chars)) != -1)
      {
        buffer.append(chars, 0, read); 
      }

      return buffer.toString();
    }
		finally 
		{
			try 
			{
				if (reader != null)
        {
					reader.close();
        }
			}
			catch (IOException e)
			{
			}
    }
  }

  public Object ReadJson(String uri) throws IOException
  {
    String content = ReadString(uri);
    Object json = JSONValue.parse(content);

    if (json == null)
    {
      throw new IOException();
    }
    else
    {
      return json;
    }
  }

  public BufferedImage ReadImage(String uri) throws IOException
  {
    return ImageIO.read(OpenStream(uri));
  }

  public String GetName()
  {
    return name_;
  }

  public void SetName(String name)
  {
    name_ = name;
  }

  public JSONObject Serialize()
  {
    JSONObject json = new JSONObject();
    json.put("Name", name_);
    json.put("Url", baseUrl_);

    if (authentication_ != null)
    {
      json.put("Authentication", authentication_);
    }

    return json;
  }

  public static OrthancConnection Unserialize(JSONObject json)
  {
    OrthancConnection c = new OrthancConnection();
    c.SetName((String) json.get("Name"));
    c.SetBaseUrl((String) json.get("Url"));

    if (json.containsKey("Authentication"))
    {
      c.authentication_ = (String) json.get("Authentication");
    }
    else
    {
      c.authentication_ = null;
    }

    return c;
  }
}


com/orthancserver/OrthancConfigurationDialog.java
/**
 * Orthanc - A Lightweight, RESTful DICOM Store
 * Copyright (C) 2012-2014 Medical Physics Department, CHU of Liege,
 * Belgium
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 **/


package com.orthancserver;

import java.awt.BorderLayout;
import java.awt.Canvas;
import java.awt.FlowLayout;
import java.awt.GridLayout;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.io.IOException;
import javax.swing.BoxLayout;
import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JLabel;
import javax.swing.JOptionPane;
import javax.swing.JPanel;
import javax.swing.JPasswordField;
import javax.swing.JScrollPane;
import javax.swing.JSplitPane;
import javax.swing.JTextField;
import javax.swing.JTree;
import javax.swing.SwingWorker;
import javax.swing.border.EmptyBorder;

import org.json.simple.JSONObject;


public class OrthancConfigurationDialog extends JDialog 
{
  private JTextField name_ = new JTextField("Orthanc");
  private JTextField url_ = new JTextField("http://localhost:8042/");
  private JTextField username_ = new JTextField("");
  private JTextField password_ = new JPasswordField("");
  private boolean isSuccess_;

  private OrthancConnection CreateConnection()
  {
    OrthancConnection orthanc = new OrthancConnection();
    orthanc.SetName(name_.getText());
    orthanc.SetBaseUrl(url_.getText());

    if (username_.getText().length() != 0)
    {
      orthanc.SetCredentials(username_.getText(), password_.getText());
    }

    return orthanc;
  }

  public OrthancConfigurationDialog()
  {
    final JPanel contentPanel = new JPanel();
		contentPanel.setBorder(new EmptyBorder(20, 5, 5, 5));
		contentPanel.setLayout(new GridLayout2(0, 2, 20, 5));

    JLabel label = new JLabel("Name:");
    label.setHorizontalAlignment(JLabel.RIGHT);
    contentPanel.add(label);
    contentPanel.add(name_);

    label = new JLabel("URL:");
    label.setHorizontalAlignment(JLabel.RIGHT);
    contentPanel.add(label);
    contentPanel.add(url_);

    label = new JLabel("Username:");
    label.setHorizontalAlignment(JLabel.RIGHT);
    contentPanel.add(label);
    contentPanel.add(username_);

    label = new JLabel("Password:");
    label.setHorizontalAlignment(JLabel.RIGHT);
    contentPanel.add(label);
    contentPanel.add(password_);

    getContentPane().setLayout(new BorderLayout());
    getContentPane().add(contentPanel, BorderLayout.NORTH);

    JPanel buttonPane = new JPanel();
    buttonPane.setLayout(new FlowLayout(FlowLayout.RIGHT));
    getContentPane().add(buttonPane, BorderLayout.SOUTH);
		{
			{
        JButton test = new JButton("Test connection");
				test.addActionListener(new ActionListener() {
          public void actionPerformed(ActionEvent arg) {
            OrthancConnection orthanc = CreateConnection();
            try
            {
              JSONObject system = (JSONObject) orthanc.ReadJson("system");
              JOptionPane.showMessageDialog(null, "Successfully connected to this Orthanc server " +
                                            "(version: " + (String) system.get("Version") + ")!", 
                                            "Success", JOptionPane.INFORMATION_MESSAGE);
            }
            catch (IOException e)
            {
              JOptionPane.showMessageDialog(null, "Cannot connect to this Orthanc server!", 
                                            "Error", JOptionPane.ERROR_MESSAGE);
            }
          }
        });
				buttonPane.add(test);
			}
			{
        JButton okButton = new JButton("Add");
				okButton.addActionListener(new ActionListener() {
          public void actionPerformed(ActionEvent arg) {
            isSuccess_ = true;
            setVisible(false);
          }
        });
				buttonPane.add(okButton);
				getRootPane().setDefaultButton(okButton);
			}
			{
				JButton cancelButton = new JButton("Cancel");
				cancelButton.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent arg) {
            setVisible(false);
					}
				});
				buttonPane.add(cancelButton);
			}
		}

    setUndecorated(false);
    setSize(500,500);
    setTitle("Add new server");
    setModal(true);
  }



  public OrthancConnection ShowModal()
  {
    isSuccess_ = false;
    setVisible(true);
    if (isSuccess_)
    {
      return CreateConnection();
    }
    else
    {
      return null;
    }
  }

}
